File: ztm_master_coding_interview_course.py/linked_lists/linked_list.py
from typing import Any

class Node:
    """ Create a separated class to store data into nodes. """
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None


    def prepend(self, data: Any) -> None:
        """ Add nodes to head. Time complexity O(1). """
        if self.head is None:
            add_node = Node(data=data)
            self.head = add_node
        else:
            add_node = Node(data=data)
            add_node.next = self.head
            self.head = add_node

    
    def print_out_linked_list(self) -> str | False:
        """ 
        Print out all nodes in the linked list. 
        Last node references to a None value.
        Lastly, if Linked List is empty, return False value.
        """
        if self.head is None:
            print("Sorry, linked list does not has info :(")
            return False

        while self.head is not None:
            print(self.head.data, ' --> ', end=' ')
            self.head = self.head.next
            if self.head is None:
                print(None, end=' ')
        
        print('\nExecution completed successfully :D')



if __name__ == '__main__':
    my_linked_list = LinkedList() # Instance of LinkedList class.
    
    # Add nodes to head.
    my_linked_list.prepend(10)
    my_linked_list.prepend(5)
    my_linked_list.prepend(15)
    my_linked_list.prepend("Tiburoncin uahah")
    
    # Print out linked list.
    my_linked_list.print_out_linked_list()

File: ztm_master_coding_interview_course.py/linked_lists/linked_list.js
class Node {
    constructor (data) {
        this.data = data
        this.next = null
    }
}


class LinkedList {
    constructor (data) {
        this.head = {
            data: data,
            next: null
        }
        this.tail = this.head
    }


    prepend (data) {
        /* Add nodes to linked list head. */
        if (this.head === null) {
            let addNode = new Node(data)
            this.head = addNode
        }
        else {
            const newNode = new Node(data)
            newNode.next = this.head
            this.head = newNode
        }
    }


    appendNodes (data) {
        /* Insert data at linked list's tail. Time complexity O(1). */
        if (this.head === null) {
            this.prepend(data)
        }
        else {
            const newNode = new Node(data)
            this.tail.next = newNode
            this.tail = newNode
        }
    }


    insertNodeAt (index, data) {
        /* Insert new data at index given. Time complexity O(n). */

        if (index < 0) {
            throw new Error('Invalid index...')
        } else if (index === 0) {
            this.prepend(data)
        }

        let count = 0
        let iterator = this.head

        while (iterator !== null) {
            if (count === index - 1) {
                const newNode = new Node(data) // Create new node.
                newNode.next = iterator.next // Set the new node's next pointer to the next node of the current node.
                iterator.next = newNode // Set the current node's next pointer to the new node.
                return
            }
            count++
            iterator = iterator.next
        }
        throw new Error('Linked list out of bounds...')
    }


    removeNode (index) {
        /* Remove node at index given. Time complexity O(n) if the index is greater than zero, if it is equals to zero time complexity would be O(1). */
        if (index < 0) {
            throw new Error('Invalid index...')
        } else if (index === 0) {
            this.head = this.head.next
            return
        }

        let count = 0
        let iterator = this.head

        while (iterator !== null) {
            if (count === index - 1) {
                // If you want to checkout which item will be removed, uncomment the following line:
                //console.log("Node to remove ->", iterator.next.data);
                iterator.next = iterator.next.next
                return
            }
            count++
            iterator = iterator.next
        }
        throw new Error('Linked list out of bounds...')
    }


    printOutLinkedList () {
        let iterator = this.head
        const llArray = [];

        while (iterator !== null) {
            llArray.push(iterator.data)
            iterator = iterator.next
        }
        return llArray
    }
}


// Example usage.
const MyLinkedList = new LinkedList(10)

MyLinkedList.prepend(5)
MyLinkedList.prepend(115)
MyLinkedList.appendNodes(15)
MyLinkedList.insertNodeAt(2, "Holi")
MyLinkedList.removeNode(4)

console.log(MyLinkedList.printOutLinkedList())
